File: apps/meridian-ui/src/api/telemetry.ts
import { SpanStatusCode, trace } from "@opentelemetry/api";
import type { Span } from "@opentelemetry/api";
import type { AuditLog } from "../workspace/audit.js";
import type { AuditLogEntry } from "../workspace/types.js";
import { ApiError } from "./client.js";

const TRACER_NAME = "meridian.api";
const TRACER_VERSION = "0.1.0";

export function getTracer() {
  return trace.getTracer(TRACER_NAME, TRACER_VERSION);
}

export interface ApiInvocationEvent {
  readonly name: string;
  readonly operation: string;
  readonly timestamp: string;
  readonly sessionId?: string;
}

export function recordApiInvocationEvent(span: Span, event: ApiInvocationEvent): void {
  span.addEvent("api.invocation", {
    "event.name": event.name,
    "api.operation": event.operation,
    "event.timestamp": event.timestamp,
    ...(event.sessionId ? { "session.id": event.sessionId } : {}),
  });
}

export function recordApiFailure(
  span: Span,
  err: unknown,
  auditLog: AuditLog,
  context: { operation: string; sessionId?: string },
): void {
  const message = err instanceof Error ? err.message : String(err);
  const code = err instanceof ApiError ? String(err.status) : "API_ERROR";
  span.setStatus({ code: SpanStatusCode.ERROR, message });
  if (err instanceof Error) span.recordException(err);

  const entry: AuditLogEntry = {
    level: "error",
    event: `api.${context.operation}.failed`,
    sessionId: context.sessionId ?? "",
    timestamp: new Date().toISOString(),
    detail: { code, message },
  };
  auditLog.write(entry);
}

File: packages/api-models/src/api_models/_telemetry.py
from __future__ import annotations

from opentelemetry import trace
from opentelemetry.trace import Span, Status, StatusCode

from ._version import API_MODELS_VERSION

_TRACER_NAME = "meridian.api-models"


def get_tracer() -> trace.Tracer:
    return trace.get_tracer(_TRACER_NAME, API_MODELS_VERSION)


def record_list_event(span: Span, *, count: int) -> None:
    """Attach a structured models.list.invocation event to the active span."""
    span.add_event("models.list.invocation", {"count": count})


def record_failure(span: Span, error: Exception, *, operation: str) -> None:
    """Set span status to ERROR, add a models.error event, and record the exception."""
    span.set_status(Status(StatusCode.ERROR, str(error)))
    span.add_event(
        "models.error",
        {
            "operation": operation,
            "error.type": type(error).__name__,
            "error.message": str(error),
        },
    )
    span.record_exception(error)
